openmc/tallies.py

```diff
@@ -0,0 +1,3204 @@
+
+        # If all filters, nuclides and scores match then tallies are mergeable
+        if equal_filters and equal_nuclides and equal_scores:
+            return True
+
+        # Variables to indicate matching filter bins, nuclides and scores
+        merge_filters = self._can_merge_filters(other)
+        merge_nuclides = self._can_merge_nuclides(other)
+        merge_scores = self._can_merge_scores(other)
+        mergeability = [merge_filters, merge_nuclides, merge_scores]
+
+        if not all(mergeability):
+            return False
+
+        # If the tally results have been read from the statepoint, we can only
+        # at least two of filters, nuclides and scores must match
+        elif self._results_read and sum(equality) < 2:
+            return False
+        else:
+            return True
+
+    def merge(self, other):
+        """Merge another tally with this one
+
+        If results have been loaded from a statepoint, then tallies are only
+        mergeable along one and only one of filter bins, nuclides or scores.
+
+        Parameters
+        ----------
+        other : openmc.Tally
+            Tally to merge with this one
+
+        Returns
+        -------
+        merged_tally : openmc.Tally
+            Merged tallies
+
+        """
+
+        if not self.can_merge(other):
+            msg = 'Unable to merge tally ID="{0}" with ' \
+                   '"{1}"'.format(other.id, self.id)
+            raise ValueError(msg)
+
+        # Create deep copy of tally to return as merged tally
+        merged_tally = copy.deepcopy(self)
+
+        # Differentiate Tally with a new auto-generated Tally ID
+        merged_tally.id = None
+
+        # Create deep copy of other tally to use for array concatenation
+        other_copy = copy.deepcopy(other)
+
+        # Identify if filters, nuclides and scores are mergeable and/or equal
+        merge_filters = self._can_merge_filters(other)
+        merge_nuclides = self._can_merge_nuclides(other)
+        merge_scores = self._can_merge_scores(other)
+        equal_filters = sorted(self.filters) == sorted(other.filters)
+        equal_nuclides = sorted(self.nuclides) == sorted(other.nuclides)
+        equal_scores = sorted(self.scores) == sorted(other.scores)
+
+        # If two tallies can be merged along a filter's bins
+        if merge_filters and not equal_filters:
+
+            # Search for mergeable filters
+            for i, filter1 in enumerate(self.filters):
+                for filter2 in other.filters:
+                    if filter1 != filter2 and filter1.can_merge(filter2):
+                        other_copy._swap_filters(other_copy.filters[i], filter2)
+                        merged_tally.filters[i] = filter1.merge(filter2)
+                        join_right = filter1 < filter2
+                        merge_axis = i
+                        break
+
+        # If two tallies can be merged along nuclide bins
+        if merge_nuclides and not equal_nuclides:
```

Comment:
As mentioned in the class documentation, I think a lot of this can be cleaned up by using sets.
e.g. `mergedTally.scores = mergedTally.scores | otherTally.scores`